source/fe/mapping_q_generic.cc

```diff
@@ -1527,20 +1612,71 @@ namespace internal
-                                          data.mapping_support_points[0]);
-                for (unsigned int k=1; k<data.n_shape_functions; ++k)
-                  for (unsigned int i=0; i<spacedim; ++i)
-                    result[i] += shape[k] * data.mapping_support_points[k][i];
-                quadrature_points[point] = result;
+                Assert(data.shape_info.n_q_points > 0, ExcInternalError());
+
+                const unsigned int n_shape_values = data.n_shape_functions;
+                const unsigned int n_q_points = quadrature_points.size();
+                const unsigned int vec_length = VectorizedArray<double>::n_array_elements;
+                const unsigned int n_comp = 1+ (spacedim-1)/vec_length;
+
+                Assert (data.shape_info.n_q_points == quadrature_points.size(),
+                        ExcDimensionMismatch(data.shape_info.n_q_points, quadrature_points.size()));
+
+                data.values_dofs.resize(n_comp*n_shape_values);
+                VectorizedArray<double> *values_dofs_ptr[n_comp];
+                data.values_quad.resize(n_comp*n_q_points);
+                VectorizedArray<double> *values_quad_ptr[n_comp];
+
+                for (unsigned int c=0; c<n_comp; ++c)
+                  {
+                    values_dofs_ptr[c] = &(data.values_dofs[c*n_shape_values]);
+                    values_quad_ptr[c] = &(data.values_quad[c*n_q_points]);
+                  }
+
+                for (unsigned int i=0; i<n_shape_values; ++i)
+                  for (unsigned int d=0; d<spacedim; ++d)
+                    {
+                      const unsigned int in_comp = d%vec_length;
+                      const unsigned int out_comp = d/vec_length;
+                      data.values_dofs[out_comp*n_shape_values+i][in_comp]
+                        = data.mapping_support_points[data.inverse_renumber[i]][d];
+                    }
+
+                internal::FEEvaluationImpl<internal::MatrixFreeFunctions::tensor_general, dim, -1, 0, n_comp, double>::evaluate
+                (data.shape_info, &(values_dofs_ptr[0]), &(values_quad_ptr[0]), nullptr, nullptr,
+                 &(data.scratch[0]), true, false, false);
+
+                for (unsigned int out_comp=0; out_comp<n_comp-1; ++out_comp)
+                  for (unsigned int i=0; i<n_q_points; ++i)
+                    for (unsigned int in_comp=0; in_comp<vec_length; ++in_comp)
+                      quadrature_points[i][out_comp*vec_length+in_comp]
+                        = data.values_quad[out_comp*n_q_points+i][in_comp];
+                // Treat the last component special as it might not be full.
+                for (unsigned int i=0; i<n_q_points; ++i)
+                  for (unsigned int in_comp=0; in_comp<spacedim-(n_comp-1)*vec_length; ++in_comp)
+                    quadrature_points[i][(n_comp-1)*vec_length+in_comp]
+                      = data.values_quad[(n_comp-1)*n_q_points+i][in_comp];
```

Comment:
I don't understand the point of using a separate loop here. All the bounds should be known at compile time, so checking for `in_comp<vec_length && in_comp<spacedim-(n_comp-1)*vec_length` should work, shouldn't it?